### PR TITLE
Optimize code printing

### LINKED ISSUE #2438: Code highlighting isn't visible for printing
### Please confirm that you have searched existing issues in the repo

Yes, I have searched the existing issues

### Any related issues?

_No response_

### Tell us about your environment

macOS with latest Chrome or Safari

### MarkBind version

5.3.0

### Describe the bug and the steps to reproduce it

Line highlighting doesn't seem to be visible when printing. From `markbind.css`, it appears that the background color for code highlighting is intended for Lavander.

https://github.com/MarkBind/markbind/blob/e1ba80eebe18375e4a2d6f69d845bd418ee27d94/packages/core-web/src/styles/markbind.css#L67-L82

Web view:
Screenshot 2024-02-26 at 14 14 32
Print view:
Screenshot 2024-02-26 at 14 14 54

### Expected behavior

The highlighting should be visible.

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by yucheng11122017**
**What is the purpose of this pull request?**

- [ ] Documentation update
- [X] Bug fix
- [ ] Feature addition or enhancement
- [ ] Code maintenance
- [ ] DevOps
- [ ] Improve developer experience
- [ ] Others, please explain:

Resolves #2438 

**Overview of changes:**
Add CSS which 
- Makes code light theme automatically when printing 
- Removes code wrap and code copy buttons when printing 
- Include code highlighting when printing 

**Anything you'd like to highlight/discuss:**
If anyone can help to test on Mac, that will be great. 

There seems to be some issue with printing on Edge.. the CSS isnt applied on the preview even though the emulator reflects the changes.. 
TLDR: Don;t use Edge to print

**Testing instructions:**
Make code theme dark 

```js {start-from=6 .line-numbers highlight-lines="7, 9"}
function add(a, b) {
    const sum = a + b;
    console.log(`${a} + ${b} = ${sum}`);
    return sum;
}
```

Notice that code wrap and code copy buttons are gone
Code highlight is printed 
Code theme is light

**Proposed commit message: (wrap lines at 72 characters)**
Optimize code blocks for printing

---

**Checklist:** :ballot_box_with_check:

- [X] Updated the documentation for feature additions and enhancements
- [ ] Added tests for bug fixes or features
- [X] Linked all related issues
- [X] No unrelated changes 

---

**Reviewer checklist:**

Indicate the [SEMVER](https://semver.org/) impact of the PR:
- [ ] Major (when you make incompatible API changes)
- [ ] Minor (when you add functionality in a backward compatible manner)
- [x] Patch (when you make backward compatible bug fixes)

At the end of the review, please label the PR with the appropriate label: `r.Major`, `r.Minor`, `r.Patch`.

Breaking change release note preparation (if applicable):
- To be included in the release note for any feature that is made obsolete/breaking

> Give a brief explanation note about:
>   - what was the old feature that was made obsolete
>   - any replacement feature (if any), and
>   - how the author should modify his website to migrate from the old feature to the replacement feature (if possible).

### CODE REVIEW

**Comment by itsyme** [comment]
Tested it out on MacOS on Chrome and Arc browsers, the making code theme light automatically and the removal or buttons seems to work, but the code highlighting doesn't seem to work for me.
Screenshot 2024-03-08 at 11 21 24 PM

**Comment by Tim-Siu** [comment]
I also do not see code highlighting in printing.

Another thing that can be considered is that when we are in the dark theme and print, the background does switch to white, but the highlight color remains to be that of the dark theme and the color is a bit muted.

File: packages/core-web/src/styles/markbind.css
Comment (Tim-Siu) [inline packages/core-web/src/styles/markbind.css:1]: It may be necessary to override the default code highlight color scheme for dark themes when printing. Otherwise, the color seems to be slightly muted.

**Review by Tim-Siu — COMMENTED** [review]
I am also not able to observe code highlighting.

Another interesting problem is that although we automatically switched to the light theme in printing, the code highlighting color isn't updated and it is comparatively muted.

Above: light colour theme highlight scheme:
Screenshot 2024-03-09 at 12 46 39

Below: dark colour theme highlight scheme:
Screenshot 2024-03-09 at 12 47 54

**Comment by yucheng11122017** [comment]
Hi @itsyme and @Tim-Siu thank you for the work on reviewing! I have made some changes accordingly. Do you mind helping me check on Mac again? 

I'm not sure what's with the highlighting not appearing on the print preview despite it being on the emulator.. Could you help me check if thats the case on Safari and etc as well?

**Comment by Tim-Siu** [comment]
> Hi @itsyme and @Tim-Siu thank you for the work on reviewing! I have made some changes accordingly. Do you mind helping me check on Mac again?
> 
> I'm not sure what's with the highlighting not appearing on the print preview despite it being on the emulator.. Could you help me check if thats the case on Safari and etc as well?

Hi @yucheng11122017 ,
The highlighting is still not working on Mac though.

**Comment by yucheng11122017** [comment]
> > Hi @itsyme and @Tim-Siu thank you for the work on reviewing! I have made some changes accordingly. Do you mind helping me check on Mac again?
> > I'm not sure what's with the highlighting not appearing on the print preview despite it being on the emulator.. Could you help me check if thats the case on Safari and etc as well?
> 
> Hi @yucheng11122017 , The highlighting is still not working on Mac though.

On which browser? Could you check the emulator as well?

**Comment by Tim-Siu** [comment]
> > > Hi @itsyme and @Tim-Siu thank you for the work on reviewing! I have made some changes accordingly. Do you mind helping me check on Mac again?
> > > I'm not sure what's with the highlighting not appearing on the print preview despite it being on the emulator.. Could you help me check if thats the case on Safari and etc as well?
> > 
> > 
> > Hi @yucheng11122017 , The highlighting is still not working on Mac though.
> 
> On which browser? Could you check the emulator as well?

Hi @yucheng11122017 

May I know what is the emulator? Do you mean the preview before printing?

The code highlighting isn't working on either Safari or Chrome in both preview and actual printed files.

**Comment by yucheng11122017** [comment]
> > > > Hi @itsyme and @Tim-Siu thank you for the work on reviewing! I have made some changes accordingly. Do you mind helping me check on Mac again?
> > > > I'm not sure what's with the highlighting not appearing on the print preview despite it being on the emulator.. Could you help me check if thats the case on Safari and etc as well?
> > > 
> > > 
> > > Hi @yucheng11122017 , The highlighting is still not working on Mac though.
> > 
> > 
> > On which browser? Could you check the emulator as well?
> 
> Hi @yucheng11122017
> 
> May I know what is the emulator? Do you mean the preview before printing?
> 
> The code highlighting isn't working on either Safari or Chrome in both preview and actual printed files.

You can try to open dev tools -> rendering -> print emulation. 
https://stackoverflow.com/questions/9540990/using-chromes-element-inspector-in-print-preview-mode

**Comment by Tim-Siu** [comment]
> > > > > Hi @itsyme and @Tim-Siu thank you for the work on reviewing! I have made some changes accordingly. Do you mind helping me check on Mac again?
> > > > > I'm not sure what's with the highlighting not appearing on the print preview despite it being on the emulator.. Could you help me check if thats the case on Safari and etc as well?
> > > > 
> > > > 
> > > > Hi @yucheng11122017 , The highlighting is still not working on Mac though.
> > > 
> > > 
> > > On which browser? Could you check the emulator as well?
> > 
> > 
> > Hi @yucheng11122017
> > May I know what is the emulator? Do you mean the preview before printing?
> > The code highlighting isn't working on either Safari or Chrome in both preview and actual printed files.
> 
> You can try to open dev tools -> rendering -> print emulation. https://stackoverflow.com/questions/9540990/using-chromes-element-inspector-in-print-preview-mode

Hi @yucheng11122017 ,
Screenshot 2024-03-09 at 17 57 21
This is what I get from the emulator. We do see highlight colour here.

**Comment by itsyme** [comment]
I tried it on the emulator and it works for me as well however I'm not sure if it working on the emulator may be a reliable indicator of it working when printing in the physical form (https://stackoverflow.com/questions/26123622/print-preview-not-the-same-as-print-emulation-in-chrome) as when I downloaded to PDF (below) the highlights were still not appearing for me. To be 100% sure I can try printing a physical copy out to check if the highlights work.
Screenshot 2024-03-10 at 8 03 53 AM

**Comment by yucheng11122017** [comment]
> I tried it on the emulator and it works for me as well however I'm not sure if it working on the emulator may be a reliable indicator of it working when printing in the physical form (https://stackoverflow.com/questions/26123622/print-preview-not-the-same-as-print-emulation-in-chrome) as when I downloaded to PDF (below) the highlights were still not appearing for me. To be 100% sure I can try printing a physical copy out to check if the highlights work. Screenshot 2024-03-10 at 8 03 53 AM

I think its cool @itsyme. I tried printing out that day but it doesn't work as well.. Let me see if there's a workaround

**Comment by tlylt** [comment]
> I think its cool @itsyme. I tried printing out that day but it doesn't work as well.. Let me see if there's a workaround

I believe the generated pdf would be the source of truth, if it has highlight, then the printed version will do too.

**Comment by itsyme** [comment]
Perhaps it would be good to include the findings in the user guide about which browsers are the most optimal for printing so that users interested in printing can refer to it!

**Comment by Tim-Siu** [comment]
HI @yucheng11122017 ,

I found this [discussion](https://stackoverflow.com/questions/2392366/print-background-colours-in-chrome) to be relevant. It appears that adding `-webkit-print-color-adjust:exact;` in the element css will force print the background colour.

**Example:**
```css
    .hljs span.highlighted{
        -webkit-print-color-adjust:exact;
        background:#e6e6fa !important;
    }
```

**Output:**

Screenshot 2024-03-15 at 20 05 57
<br/>

Alternatively, users can also choose to print background colour themselves in Chrome settings under more settings -> options.
Screenshot 2024-03-15 at 20 04 10

You may refer to my [branch](https://github.com/Tim-Siu/markbind/tree/fix-print-code-highlight) which seems to have resolved the issue!

**Comment by kaixin-hc** [comment]
might want to make sure it "works" on safari and firefox also - i believe they use the non-prefixed css ([stack overflow link](https://stackoverflow.com/questions/35625178/what-is-the-alternate-for-webkit-print-color-adjust-in-firefox-and-ie))

**Comment by Tim-Siu** [comment]
> might want to make sure it "works" on safari and firefox also - i believe they use the non-prefixed css ([stack overflow link](https://stackoverflow.com/questions/35625178/what-is-the-alternate-for-webkit-print-color-adjust-in-firefox-and-ie))

Hi @kaixin-hc ,

This approach works for Safari as well. (Maybe because Safari is also based on WebKit) It doesn't work for Firefox though. From my observations, there seem to have been many other bugs when printing with Firefox. I am not sure if it is necessary to fix them all.


Imag: printing result with firefox and current master branch
Screenshot 2024-03-18 at 12 09 27

**Comment by kaixin-hc** [comment]
Great investigation! @yucheng11122017 - perhaps you can implement the fix @Tim-Siu suggests, and then include a documentation note that there are known errors with printing on Edge and Firefox? 

In terms of if we should fix it for Firefox/Edge, though it would be nice to figure out why and support it, if printing is OK on Chrome and Safari I think it is OK to deprioritise this (since a brief google search claims 70% of people are on Chrome or safari, with only 7% on Edge or Firefox)

**Comment by yucheng11122017** [comment]
Apologies for the late response! I missed this in the slew of emails.. 
Thanks so much for the investigative work @Tim-Siu! It's really great and it seems to be working on Edge on Windows 

image

I will push the fix and also add a caveat about Firefox in that case.

File: docs/userGuide/syntax/code.md [hunk 1]
```diff
@@ -281,7 +281,17 @@ A _wrap text_ button can be added to code blocks using the `codeBlockWrapButtons
...
-Markbind enhances the readability of your code blocks for printing by applying soft wrapping, ensuring code doesn't get cut off. Additionally, line numbers are added to maintain context when wrapping occurs.
+Markbind enhances the readability of your code blocks for printing by 
+
+- Applying soft wrapping, ensuring code doesn't get cut off
+- Adding line numbers to maintain context when wrapping occurs
+- Automatically changing dark code theme to light
+- Removing [codeBlockCopyButtons]({{ baseUrl }}/userGuide/formattingContents.html#copy-button) and [codeBlockWrapButtons]({{baseUrl}}/userGuide/formattingContents.html#wrap-text-button)
+
+<box type="warning" seamless> 
+    There are some issues with printing in Firefox. Please use other browsers if these issues persist.
```
Comment (itsyme) [inline docs/userGuide/syntax/code.md:292]: I feel it could be better to specify suggested browser(s) e.g. Chrome as it may give the user more certainty when looking for an alternative browser if they are using Firefox

**Review by itsyme — APPROVED** [review]
LGTM! Just a small nit in the wording of the documentation.

File: docs/userGuide/syntax/code.md [hunk 2]
```diff
@@ -281,7 +281,17 @@ A _wrap text_ button can be added to code blocks using the `codeBlockWrapButtons
 ##### Printing optimization
 
 <div id="code-print-optimization">
-Markbind enhances the readability of your code blocks for printing by applying soft wrapping, ensuring code doesn't get cut off. Additionally, line numbers are added to maintain context when wrapping occurs.
+Markbind enhances the readability of your code blocks for printing by 
+
+- Applying soft wrapping, ensuring code doesn't get cut off
+- Adding line numbers to maintain context when wrapping occurs
+- Automatically changing dark code theme to light
+- Removing [codeBlockCopyButtons]({{ baseUrl }}/userGuide/formattingContents.html#copy-button) and [codeBlockWrapButtons]({{baseUrl}}/userGuide/formattingContents.html#wrap-text-button)
```
Comment (itsyme) [inline docs/userGuide/syntax/code.md:289]:
```suggestion
- Removing [codeBlockCopyButtons]({{ baseUrl }}/userGuide/formattingContents.html#copy-button) and [codeBlockWrapButtons]({{ baseUrl }}/userGuide/formattingContents.html#wrap-text-button)
```

Comment (yucheng11122017) [reply docs/userGuide/syntax/code.md]: Thanks for comments @itsyme! I think it should be the other way around actually - I removed the spaces. Otherwise the formatting doesn't show the link
